auth/postgres/repo.go

```diff
@@ -0,0 +1,465 @@
+
+	upm := dbPatPagemeta{
+		User:      userID,
+		ID:        patID,
+		UpdatedAt: time.Now(),
+		ExpiresAt: expiryAt,
+		Secret:    tokenHash,
+	}
+
+	rows, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return auth.PAT{}, errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	var pat auth.PAT
+	if err := rows.StructScan(&pat); err != nil {
+		return auth.PAT{}, errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return pat, nil
+}
+
+func (pr *patRepo) Revoke(ctx context.Context, userID, patID string) error {
+	q := `
+		UPDATE pats 
+		SET revoked = true, revoked_at = :revoked_at
+		WHERE user_id = :user_id AND id = :id`
+
+	upm := dbPatPagemeta{
+		User:      userID,
+		ID:        patID,
+		ExpiresAt: time.Now(),
+	}
+
+	_, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) Reactivate(ctx context.Context, userID, patID string) error {
+	q := `
+		UPDATE pats 
+		SET revoked = false, revoked_at = NULL
+		WHERE user_id = :user_id AND id = :id`
+
+	upm := dbPatPagemeta{
+		User: userID,
+		ID:   patID,
+	}
+
+	_, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) Remove(ctx context.Context, userID, patID string) error {
+	q := `DELETE FROM pats WHERE user_id = :user_id AND id = :id`
+	upm := dbPatPagemeta{
+		User: userID,
+		ID:   patID,
+	}
+
+	_, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrRemoveEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) AddScopeEntry(ctx context.Context, userID string, scopes []auth.Scope) error {
+	scs := toDBScope(scopes)
+	_, err := pr.db.NamedQueryContext(ctx, saveScopeQuery, scs)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrUpdateEntity, err)
+	}
+
+	if err := pr.cache.Save(ctx, scopes); err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) RemoveScopeEntry(ctx context.Context, userID string, scopes []auth.Scope) error {
+	dbscopes := toDBScope(scopes)
+
+	_, err := pr.db.NamedQueryContext(ctx, deleteScopesQuery, dbscopes)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	if err := pr.cache.Remove(ctx, scopes); err != nil {
+		return errors.Wrap(repoerr.ErrRemoveEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) CheckScopeEntry(ctx context.Context, userID, patID string, entityType auth.EntityType, optionalDomainID string, operation auth.Operation, entityID string) error {
+	authorized, err := pr.cache.CheckScope(ctx, patID, optionalDomainID, entityType, operation, entityID)
+	if err == nil && authorized {
+		return nil
+	}
+
+	scopes, err := pr.retrieveScopeFromDB(ctx, patID)
+	if err != nil {
+		return err
+	}
```

Comment:
Can we have a query which retrieve only the requested scope from database 
like select from pat_scope table only the row which match the request ?